### ISSUE TITLE
Help with performance issue

### ISSUE THREAD

**Issue by peter1745** [jrouwe/JoltPhysics#265]
Hello! I just want to start by saying thank you for making Jolt! I've been super excited to implement it into the engine I'm working on, which currently uses PhysX. I'm really liking the API so far, and I love that it takes advantage of multiple cores. But I do have a pretty serious performance issue right now, which I'm not having with PhysX, my current assumption is that I'm using the API incorrectly so I figured I'd open an issue here to a) make sure it's me and not a limitation of Jolt, and b) figure out what I'm doing wrong (and hopefully learn something in the process)

So I'm trying to simulate 5625 dynamic cubes (in order to really stress test Jolt vs. PhysX), and here's what I'm currently doing:
* Creating the cubes, and adding them in bulk
* Optimizing the broad phase immediately after
* Simulating using multiple collision sub steps (currently 2 due to the low frame time), and 1 integration sub step

The image below is a screenshot of my Optick profiling session (I've integrated Jolt's external profiler system with Optick), and Jolt used in total 16 threads.
![image](https://user-images.githubusercontent.com/7275084/194939099-1420b2e5-2c18-4c43-b828-d1b581434b7f.png)

I noticed that the velocity constraint solver job took roughly 6ms, and the position constraints solver takes roughly 20ms each frame. And the "find collisions" job takes roughly 8ms, *per thread*.

Now of course 5625 dynamic objects is a **lot** to process, but when we ran the exact same test using PhysX each simulation frame took 13ms total, and we're not using the multi-threading feature of PhysX, so it's computing everything on the main thread.

Now naturally there will be cases where Jolt is faster, and cases where PhysX is faster, but since there's this big of a difference, and considering the fact that Jolt is utilizing 16 threads whereas PhysX is utilizing 1 I'm 99.9% sure I'm doing something wrong.

Now I will say that our Physics system is running on the main thread, so that will of course help things, but I still don't think it should be this slow despite that.

Here's some more info about my setup:
1. I'm running on an AMD Ryzen 7 5700G CPU
2. The cubes only have box colliders, and the slow performance happens before they start colliding with anything
3. The cubes each have a mass of 10kg
4. I've specified that there can be a max of 5700 bodies in the scene, 11400 body mutexes, 28500 body pairs and 28500 contact constraints.
5. I'm using the `TempAllocatorImpl` allocator with 300MB pre-allocated memory
6. 2048 physics jobs and 8 barriers in total

Now, I've pretty much eliminated the possibility that my existing usage of the API is causing the slow down (and the profiling confirmed it for me) so at this point I'm pretty much assuming I'm either forgetting to do something, doing something wrong, or it could be that I haven't tweaked the max number of jobs and barriers.

I'd be happy to show you my code, but I figured I'd start with this 🙂 Either way, thank you for creating Jolt! Despite this issue (which is 99.9% *my* fault) I still prefer the API to something like PhysX 🙂

**Comment by peter1745** [comment]
Alright just wanted to say I just tried doubling the max physics jobs **and** the barriers, so 4096 jobs and 16 barriers, and that helped, but still looking at like 17ms for the position constraint solving, and 5ms for the velocity constraint solving, not sure if I'm just underestimating my numbers here, but yeah that's what I found

**Comment by peter1745** [comment]
Yes you are correct, there were objects colliding which I didn't realize, although I am using the default mNumPositionSteps of 2, in fact I haven't actually changed the PhysicsSettings at all for the time

**Comment by peter1745** [comment]
I did another profiling pass, this time making sure there weren't any collision for the first few seconds, and the velocity constraint solver seems to basically be taking 0ms now, but the position constraint solver still takes 20ms for some reason

**Comment by jrouwe** [comment]
Can you maybe send a screenshot that's more zoomed in on SolvePositionConstraints (on a typical 'Island') so that the sub-bars can be read? Take a look at how much each of these take in total for the entire frame. A function level profile would be nice too.

If you look at PhysicsSystem::JobSolveVelocityConstraints you'll see that it starts almost exactly the same as PhysicsSystem::JobSolvePositionConstraints by walking over the simulation islands and not doing anything if there are no contacts or constraints. The thing that PhysicsSystem::JobSolvePositionConstraints does extra at the end is to check if objects can go to sleep and to update the broad phase:

![image](https://user-images.githubusercontent.com/1621693/195183051-0f4c1cf3-11a0-4708-9f42-9ff9793d4373.png)

Assuming that there are indeed no collisions or constraints, ContactConstraintManager::SolvePositionConstraints should be really small. I can't imagine that 'check sleeping' is going to be the culprit, so that leaves notifying the broad phase. Did you set your broad phase up so that there are at least 2 trees: 1 for static objects, 1 for moving?

Have you tried replicating the scene in the SamplesApp (if you do and the problem persists then it will be really easy for me to debug)? Note that you can make a snapshot of the world with code similar to what is found in SamplesApp::TakeSnapshot, you can directly load that in the SamplesApp.

**Comment by peter1745** [comment]
Alright, so I tried replicating it in the SamplesApp, and the simulation runs just fine. So at this point it's clear that I'm doing something incorrectly, now the question is *what*. I compared the SamplesApp code to the code I have and it's pretty much identical.
I even went in and corrected some of the differences just to make sure, and it's still slow. So I guess it's back to profiling, I'll see if I can get a function-level profile, and I'll send you a better screenshot of the previous result (I could also send you the Optick files if you want).

I was considering the possibility that I had configured the library incorrectly (I wrote a premake file for it to integrate it with our engine), but I've gone in and double checked that I had the correct preprocessor definitions enabled, and I did so.

**Comment by peter1745** [comment]
Alright never mind. One of the changes I made to the code or the preprocessor statements worked, the slowdown I was seeing after making those changes were caused by my profiling integration.

But I'll still investigate which change fixed the *actual* performance issue, because before I added the Optick - Jolt interop we did notice that the Jolt simulation was slower than PhysX, so yeah, one of my changes fixed it, I'll update this issue once I've figured out which thing fixed it

**Comment by peter1745** [comment]
Alright this is really weird. I undid all the changes I made, and it's still performing decently. The weird thing is that when I demonstrated the *exact* code I've got now to my boss the performance was terrible compared to PhysX, and now it's perfectly fine.

I *was* missing some pre-processor statements, and when I undid those it still performed fine. So yeah, unsure what happened.

I'll probably close this issue for now, since the issue apparently resolved itself, somehow. But I'll keep an eye out and see if I figure something out, in which case I'll probably update this issue.

So yeah, sorry for wasting your time a bit @jrouwe, but I appreciate the support you provided, it did help me get a slightly better understanding of how Jolt works.

And once again I'm super excited for the future of Jolt, so thank you so much for making it!
